File: ios/egovframework-hyb-ios-networkapiguide/EgovPlugins/EgovFileOpenerRequestValidator.m
//
//  EgovFileOpenerRequestValidator.m
//

#import "EgovFileOpenerRequestValidator.h"
#import "EgovInterfaceRequestValidator.h"
#import "EGovComModule.h"

NSString * const EgovFileOpenerErrorSecureUrl = @"HTTPS server URL is required";
NSString * const EgovFileOpenerErrorUri = @"Download URI is not allowed";
NSString * const EgovFileOpenerErrorFilename = @"File name is invalid";
NSString * const EgovFileOpenerErrorTargetPath = @"Target path is not allowed";

static const NSInteger kMaxUriLength = 256;
static const NSInteger kMaxFilenameLength = 255;
static NSString * const kFopPrefix = @"/fop/";

static NSSet *allowedExtensions(void) {
    static NSSet *extensions;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        extensions = [NSSet setWithObjects:
                      @"txt", @"pdf", @"doc", @"docx", @"xls", @"xlsx", @"ppt", @"pptx", @"zip",
                      @"jpg", @"jpeg", @"png", @"gif", @"bmp", @"mp3", @"mp4", @"m4a", @"wav", @"3gp", @"hwp", nil];
    });
    return extensions;
}

static NSSet *dangerousExtensions(void) {
    static NSSet *extensions;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        extensions = [NSSet setWithObjects:
                      @"html", @"htm", @"js", @"jsx", @"mjs", @"jsp", @"php", @"asp", @"aspx",
                      @"sh", @"bash", @"bat", @"cmd", @"exe", @"apk", @"dex", @"so", @"dylib", @"xml", @"plist", nil];
    });
    return extensions;
}

@implementation EgovFileOpenerRequestValidator

+ (NSString *)buildDownloadUrlWithServerUrl:(NSString *)serverUrl
                                        uri:(NSString *)rawUri
                                      error:(NSError **)error {
    if (![EgovInterfaceRequestValidator isSecureServerUrl:serverUrl]) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:1
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorSecureUrl}];
        }
        return nil;
    }

    NSString *normalizedUri = [self normalizeDownloadUri:rawUri error:error];
    if (normalizedUri == nil) {
        return nil;
    }
    return [NSString stringWithFormat:@"%@%@", serverUrl, normalizedUri];
}

+ (NSString *)normalizeDownloadUri:(NSString *)rawUri error:(NSError **)error {
    if (rawUri == nil || [[rawUri stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] length] == 0) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:2
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorUri}];
        }
        return nil;
    }

    NSString *trimmed = [rawUri stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSString *path = trimmed;
    NSString *query = nil;

    NSRange queryRange = [trimmed rangeOfString:@"?"];
    if (queryRange.location != NSNotFound) {
        path = [trimmed substringToIndex:queryRange.location];
        query = [trimmed substringFromIndex:queryRange.location + 1];
    }
    NSRange fragmentRange = [path rangeOfString:@"#"];
    if (fragmentRange.location != NSNotFound) {
        path = [path substringToIndex:fragmentRange.location];
    }

    NSCharacterSet *uriPathCharset = [NSCharacterSet characterSetWithCharactersInString:
                                      @"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_./-"];
    NSCharacterSet *uriQueryCharset = [NSCharacterSet characterSetWithCharactersInString:
                                       @"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_./=&%-"];
    NSCharacterSet *invertedPath = [uriPathCharset invertedSet];
    NSCharacterSet *invertedQuery = [uriQueryCharset invertedSet];

    if (path.length > kMaxUriLength
        || [path containsString:@".."]
        || [path containsString:@"\\"]
        || [path containsString:@"%"]
        || ![path hasPrefix:@"/"]
        || ![path hasPrefix:kFopPrefix]
        || [path rangeOfCharacterFromSet:invertedPath].location != NSNotFound) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:3
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorUri}];
        }
        return nil;
    }

    if (query != nil) {
        if (query.length > kMaxUriLength || [query rangeOfCharacterFromSet:invertedQuery].location != NSNotFound) {
            if (error) {
                *error = [NSError errorWithDomain:@"EgovFileOpener" code:4
                                         userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorUri}];
            }
            return nil;
        }
        return [NSString stringWithFormat:@"%@?%@", path, query];
    }
    return path;
}

+ (NSString *)normalizeFileName:(NSString *)rawFileName error:(NSError **)error {
    if (rawFileName == nil || [[rawFileName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] length] == 0) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:5
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorFilename}];
        }
        return nil;
    }

    NSString *fileName = [[rawFileName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]]
                          stringByReplacingOccurrencesOfString:@"\\" withString:@"/"];
    NSRange slashRange = [fileName rangeOfString:@"/" options:NSBackwardsSearch];
    if (slashRange.location != NSNotFound) {
        fileName = [fileName substringFromIndex:slashRange.location + 1];
    }

    if (fileName.length > kMaxFilenameLength
        || [fileName containsString:@".."]
        || [fileName containsString:@"/"]
        || [fileName containsString:@"\\"]) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:6
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorFilename}];
        }
        return nil;
    }

    NSString *extension = [self extensionForFileName:fileName];
    if (extension.length == 0
        || [dangerousExtensions() containsObject:extension]
        || ![allowedExtensions() containsObject:extension]) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:7
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorFilename}];
        }
        return nil;
    }
    return fileName;
}

+ (NSString *)normalizeStoredFileName:(NSString *)rawStoredFileName error:(NSError **)error {
    if (rawStoredFileName == nil || [[rawStoredFileName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] length] == 0) {
        return @"";
    }

    NSString *storedFileName = [rawStoredFileName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSCharacterSet *filenameCharset = [NSCharacterSet characterSetWithCharactersInString:
                                       @"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-"];
    NSCharacterSet *invertedFilename = [filenameCharset invertedSet];

    if (storedFileName.length > kMaxFilenameLength
        || [storedFileName containsString:@".."]
        || [storedFileName containsString:@"/"]
        || [storedFileName containsString:@"\\"]
        || [storedFileName rangeOfCharacterFromSet:invertedFilename].location != NSNotFound) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:8
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorFilename}];
        }
        return nil;
    }
    return storedFileName;
}

+ (NSString *)resolveSecureTargetDirectoryForPath:(NSString *)rawTargetPath error:(NSError **)error {
    NSString *normalizedPath = [self normalizePathPrefix:rawTargetPath];
    if (normalizedPath == nil) {
        NSArray *documentPaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        normalizedPath = [[documentPaths firstObject] stringByAppendingPathComponent:@"www"];
    }

    NSString *canonicalPath = [normalizedPath stringByStandardizingPath];
    if (![self isPath:canonicalPath withinAllowedAppDirectory]) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:9
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorTargetPath}];
        }
        return nil;
    }

    NSFileManager *fileManager = [NSFileManager defaultManager];
    BOOL isDirectory = NO;
    if (![fileManager fileExistsAtPath:canonicalPath isDirectory:&isDirectory]) {
        NSError *createError = nil;
        if (![fileManager createDirectoryAtPath:canonicalPath
                     withIntermediateDirectories:YES
                                      attributes:nil
                                           error:&createError]) {
            if (error) {
                *error = createError;
            }
            return nil;
        }
        isDirectory = YES;
    } else if (!isDirectory) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:10
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorTargetPath}];
        }
        return nil;
    }
    return canonicalPath;
}

+ (NSString *)resolveSecureTargetFilePathForDirectory:(NSString *)targetDirectory
                                             fileName:(NSString *)fileName
                                                error:(NSError **)error {
    NSString *secureDirectory = [self resolveSecureTargetDirectoryForPath:targetDirectory error:error];
    if (secureDirectory == nil) {
        return nil;
    }

    NSString *targetFilePath = [[secureDirectory stringByAppendingPathComponent:fileName] stringByStandardizingPath];
    if (![self isPath:targetFilePath withinDirectory:secureDirectory]) {
        if (error) {
            *error = [NSError errorWithDomain:@"EgovFileOpener" code:11
                                     userInfo:@{NSLocalizedDescriptionKey: EgovFileOpenerErrorTargetPath}];
        }
        return nil;
    }
    return targetFilePath;
}

+ (NSString *)normalizePathPrefix:(NSString *)rawTargetPath {
    if (rawTargetPath == nil || [rawTargetPath isKindOfClass:[NSNull class]]) {
        return nil;
    }
    NSString *normalized = [rawTargetPath stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if (normalized.length == 0 || [normalized caseInsensitiveCompare:@"null"] == NSOrderedSame) {
        return nil;
    }
    if ([normalized hasPrefix:@"file://"]) {
        normalized = [normalized substringFromIndex:[@"file://" length]];
    }
    return normalized;
}

+ (BOOL)isPath:(NSString *)path withinDirectory:(NSString *)directory {
    if (path == nil || directory == nil) {
        return NO;
    }
    NSString *canonicalPath = [path stringByStandardizingPath];
    NSString *canonicalDir = [directory stringByStandardizingPath];
    if ([canonicalPath isEqualToString:canonicalDir]) {
        return YES;
    }
    return [canonicalPath hasPrefix:[canonicalDir stringByAppendingString:@"/"]];
}

+ (BOOL)isPath:(NSString *)path withinAllowedAppDirectory {
    NSArray *documentPaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSArray *cachePaths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSArray *libraryPaths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);

    for (NSString *root in documentPaths) {
        if ([self isPath:path withinDirectory:root]) {
            return YES;
        }
    }
    for (NSString *root in cachePaths) {
        if ([self isPath:path withinDirectory:root]) {
            return YES;
        }
    }
    for (NSString *root in libraryPaths) {
        if ([self isPath:path withinDirectory:root]) {
            return YES;
        }
    }
    return NO;
}

+ (NSString *)extensionForFileName:(NSString *)fileName {
    NSString *extension = [[fileName pathExtension] lowercaseString];
    return extension == nil ? @"" : extension;
}

@end

File: android-eclipse/egovframework-hyb-add-jailbreakdetectionapiguide/src/kr/go/egovframework/hyb/plugin/EgovFileOpenerRequestValidator.java
package kr.go.egovframework.hyb.plugin;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Pattern;

import org.json.JSONException;

import android.content.Context;

/**
 * Validates EgovFileOpener download requests before files are written locally.
 */
public final class EgovFileOpenerRequestValidator {

    public static final String ERROR_SECURE_URL = "HTTPS server URL is required";
    public static final String ERROR_URI = "Download URI is not allowed";
    public static final String ERROR_FILENAME = "File name is invalid";
    public static final String ERROR_TARGET_PATH = "Target path is not allowed";

    private static final int MAX_URI_LENGTH = 256;
    private static final int MAX_FILENAME_LENGTH = 255;
    private static final String FOP_PREFIX = "/fop/";

    private static final Pattern URI_PATH_PATTERN =
            Pattern.compile("^/[a-zA-Z0-9][a-zA-Z0-9_./-]*$");
    private static final Pattern URI_QUERY_PATTERN =
            Pattern.compile("^[a-zA-Z0-9_./=&%-]*$");
    private static final Pattern STORED_FILENAME_PATTERN =
            Pattern.compile("^[a-zA-Z0-9][a-zA-Z0-9._-]*$");

    private static final Set<String> ALLOWED_EXTENSIONS = new HashSet<String>(Arrays.asList(
            "txt", "pdf", "doc", "docx", "xls", "xlsx", "ppt", "pptx", "zip",
            "jpg", "jpeg", "png", "gif", "bmp", "mp3", "mp4", "m4a", "wav", "3gp", "hwp"));

    private static final Set<String> DANGEROUS_EXTENSIONS = new HashSet<String>(Arrays.asList(
            "html", "htm", "js", "jsx", "mjs", "jsp", "php", "asp", "aspx",
            "sh", "bash", "bat", "cmd", "exe", "apk", "dex", "so", "dylib", "xml", "plist"));

    private EgovFileOpenerRequestValidator() {
    }

    public static String normalizeDownloadUri(String rawUri) throws JSONException {
        if (rawUri == null || rawUri.trim().isEmpty()) {
            throw new JSONException(ERROR_URI);
        }

        String trimmed = rawUri.trim();
        String path = trimmed;
        String query = null;

        int queryIndex = trimmed.indexOf('?');
        if (queryIndex >= 0) {
            path = trimmed.substring(0, queryIndex);
            query = trimmed.substring(queryIndex + 1);
        }
        int fragmentIndex = path.indexOf('#');
        if (fragmentIndex >= 0) {
            path = path.substring(0, fragmentIndex);
        }

        if (path.length() > MAX_URI_LENGTH
                || path.contains("..")
                || path.contains("\\")
                || path.contains("%")
                || !path.startsWith("/")
                || !URI_PATH_PATTERN.matcher(path).matches()
                || !path.startsWith(FOP_PREFIX)) {
            throw new JSONException(ERROR_URI);
        }

        if (query != null) {
            if (query.length() > MAX_URI_LENGTH || !URI_QUERY_PATTERN.matcher(query).matches()) {
                throw new JSONException(ERROR_URI);
            }
            return path + "?" + query;
        }
        return path;
    }

    public static String buildDownloadUrl(Context context, String serverUrl, String rawUri)
            throws JSONException {
        if (!EgovInterfaceRequestValidator.isSecureServerUrl(context, serverUrl)) {
            throw new JSONException(ERROR_SECURE_URL);
        }
        return serverUrl + normalizeDownloadUri(rawUri);
    }

    public static String normalizeFileName(String rawFileName) throws JSONException {
        if (rawFileName == null || rawFileName.trim().isEmpty()) {
            throw new JSONException(ERROR_FILENAME);
        }

        String fileName = rawFileName.trim().replace('\\', '/');
        int slashIndex = fileName.lastIndexOf('/');
        if (slashIndex >= 0) {
            fileName = fileName.substring(slashIndex + 1);
        }

        if (fileName.isEmpty()
                || fileName.length() > MAX_FILENAME_LENGTH
                || fileName.contains("..")
                || fileName.indexOf('/') >= 0
                || fileName.indexOf('\\') >= 0
                || fileName.indexOf('\0') >= 0) {
            throw new JSONException(ERROR_FILENAME);
        }

        String extension = extractExtension(fileName);
        if (extension.isEmpty()
                || DANGEROUS_EXTENSIONS.contains(extension)
                || !ALLOWED_EXTENSIONS.contains(extension)) {
            throw new JSONException(ERROR_FILENAME);
        }
        return fileName;
    }

    public static String normalizeStoredFileName(String rawStoredFileName) throws JSONException {
        if (rawStoredFileName == null || rawStoredFileName.trim().isEmpty()) {
            return "";
        }

        String storedFileName = rawStoredFileName.trim();
        if (storedFileName.length() > MAX_FILENAME_LENGTH
                || storedFileName.contains("..")
                || storedFileName.contains("/")
                || storedFileName.contains("\\")
                || storedFileName.indexOf('\0') >= 0
                || !STORED_FILENAME_PATTERN.matcher(storedFileName).matches()) {
            throw new JSONException(ERROR_FILENAME);
        }
        return storedFileName;
    }

    public static File resolveSecureTargetFile(Context context, String rawTargetPath, String fileName)
            throws JSONException, IOException {
        File targetDirectory = resolveSecureTargetDirectory(context, rawTargetPath);
        File targetFile = new File(targetDirectory, fileName).getCanonicalFile();
        String targetDirectoryPath = targetDirectory.getCanonicalPath();
        String targetFilePath = targetFile.getPath();
        if (!targetFilePath.equals(targetDirectoryPath)
                && !targetFilePath.startsWith(targetDirectoryPath + File.separator)) {
            throw new IOException(ERROR_TARGET_PATH);
        }
        return targetFile;
    }

    public static File resolveSecureTargetDirectory(Context context, String rawTargetPath)
            throws IOException {
        String normalizedPath = normalizePathPrefix(rawTargetPath);
        if (normalizedPath == null || normalizedPath.isEmpty()) {
            normalizedPath = new File(context.getFilesDir(), "www").getPath();
        }

        File targetDirectory = new File(normalizedPath).getCanonicalFile();
        if (!isUnderAllowedAppDirectory(context, targetDirectory)) {
            throw new IOException(ERROR_TARGET_PATH);
        }
        if (!targetDirectory.exists() && !targetDirectory.mkdirs()) {
            throw new IOException("Failed to create target directory");
        }
        if (!targetDirectory.isDirectory()) {
            throw new IOException(ERROR_TARGET_PATH);
        }
        return targetDirectory;
    }

    private static String normalizePathPrefix(String rawTargetPath) {
        if (rawTargetPath == null) {
            return null;
        }
        String normalized = rawTargetPath.trim();
        if (normalized.isEmpty() || "null".equalsIgnoreCase(normalized)) {
            return null;
        }
        if (normalized.startsWith("file://")) {
            normalized = normalized.substring("file://".length());
        }
        return normalized;
    }

    private static boolean isUnderAllowedAppDirectory(Context context, File canonicalDirectory)
            throws IOException {
        String directoryPath = canonicalDirectory.getPath();
        File[] allowedRoots = getAllowedRootDirectories(context);
        for (File root : allowedRoots) {
            if (root == null) {
                continue;
            }
            String rootPath = root.getCanonicalFile().getPath();
            if (directoryPath.equals(rootPath)
                    || directoryPath.startsWith(rootPath + File.separator)) {
                return true;
            }
        }
        return false;
    }

    private static File[] getAllowedRootDirectories(Context context) throws IOException {
        File externalFilesDir = context.getExternalFilesDir(null);
        File externalCacheDir = context.getExternalCacheDir();
        return new File[] {
                context.getFilesDir(),
                context.getCacheDir(),
                externalFilesDir,
                externalCacheDir
        };
    }

    private static String extractExtension(String fileName) {
        int index = fileName.lastIndexOf('.');
        if (index < 0 || index == fileName.length() - 1) {
            return "";
        }
        return fileName.substring(index + 1).toLowerCase(Locale.ENGLISH);
    }
}
